File: app/src/main/java/Day29.java
import shared.TreeNode;

public class Day29 {

    public static boolean isIdenticalStructureAndValue(TreeNode treeNode1, TreeNode treeNode2) {

        //recursive terminal conditions
        if (treeNode1 == null && treeNode2 == null) {
            return true;  //both null, return as identical
        }

        if((treeNode1 == null || treeNode2 == null) ||
            treeNode1.getValue() != treeNode2.getValue()) {
            return false;  //they are unequal
        }

        //the current nodes are equal, so test the child node structure
        return isIdenticalStructureAndValue(treeNode1.getLeft(), treeNode2.getLeft()) &&
                isIdenticalStructureAndValue(treeNode1.getRight(), treeNode2.getRight());
    }
}

File: app/src/main/kotlin/Day29.kt
import shared.TreeNode

fun isIdenticalStructureAndValue(treeNode1: TreeNode?, treeNode2: TreeNode?): Boolean = when {
    treeNode1 == null && treeNode2 == null -> true
    treeNode1 == null || treeNode2 == null -> false
    treeNode1.value != treeNode2.value -> false

    else -> isIdenticalStructureAndValue(treeNode1.left, treeNode2.left) &&
            isIdenticalStructureAndValue(treeNode1.right, treeNode2.right)
}

